### ISSUE TITLE
Performance Investigation of tasks in the workflow

### ISSUE THREAD

**Issue by canuse** [riboviz/riboviz#325]
Investigate the performance of parallelable tasks (those can run with more than 1 thread, controlled by `num_process`) under dataset `Wallace_2020_JEC21` and dataset `Woolstenhulme_2015`.

The tasks includes
* cutAdapters (Third party tool `cutadapt`)
* hisat2rRNA (Third party tool `hisat2`)
* hisat2rORF (Third party tool `hisat2`)
* samViewSort (Third party tool `samtools`)
* bamToH5 (R script `bam_to_h5.R`)
* generateStatsFigs (R script `generate_stats_figs.R`)

I will first investigate those tasks that use third party tools as a black box, then the R scripts.

Progress
- [ ] cutAdapters
- [ ] hisat2rRNA
- [ ] hisat2rORF
- [ ] samViewSort
- [ ] bamToH5
- [ ] generateStatsFigs

**Comment by canuse** [comment]
Experiment Platform: Eddie, node3g10, node3g11, ... node3g20
Hardware: Intel Xeon Gold 6318, 20 core, 2.0GHz, 2 scokets per node, >256Gb memory per node
Software: R 3.6.3, cutadapt 2.10,  hisat2 2.1.0 (gcc 4.8.2)

**Comment by canuse** [comment]
accelerate rate of cutadapters and hisat2rRNA (real time)

![image](https://user-images.githubusercontent.com/44812363/119634105-4c100200-be0a-11eb-8e5a-c24b0f19909e.png)

the real time, user time and sys time are

![image](https://user-images.githubusercontent.com/44812363/119634246-7b267380-be0a-11eb-9245-f3483abb20b4.png)

![image](https://user-images.githubusercontent.com/44812363/119634263-811c5480-be0a-11eb-8741-b30e88dda14a.png)

![image](https://user-images.githubusercontent.com/44812363/119634301-87aacc00-be0a-11eb-96c5-90fbfeae22f6.png)

The real time is wall time, the user time is the time in user mode, the sys time is time in kernel mode (like memory allocation, I/O)
Usually, real time ~= (user time + systime) / num_process

**Comment by canuse** [comment]
hisat2ORF
summit: 8c/12c, speedup about 1.5X
weak scaling
![image](https://user-images.githubusercontent.com/44812363/120206416-15c7ed80-c223-11eb-9fbf-ef505f63f9c6.png)
![image](https://user-images.githubusercontent.com/44812363/120206464-21b3af80-c223-11eb-8545-14001e0d7afb.png)
![image](https://user-images.githubusercontent.com/44812363/120206500-2bd5ae00-c223-11eb-9d10-0e47a31fecb2.png)
![image](https://user-images.githubusercontent.com/44812363/120206524-36904300-c223-11eb-9c42-43d0cd8adf60.png)

**Comment by canuse** [comment]
samViewSort
always increasing/12~24 core
However, the maximum performance increase is less than 200%, so maybe 12 core is more cost-effective
strong scaling, however, extra time is spend in potential synchronize or CPU scheduling
![image](https://user-images.githubusercontent.com/44812363/120315401-3c922c80-c2d4-11eb-9ac1-3b5d96019afb.png)
![image](https://user-images.githubusercontent.com/44812363/120315438-49168500-c2d4-11eb-811c-bc2268873718.png)
![image](https://user-images.githubusercontent.com/44812363/120315447-4c117580-c2d4-11eb-8015-9e2d29dc5259.png)
![image](https://user-images.githubusercontent.com/44812363/120315460-4fa4fc80-c2d4-11eb-8a91-374cf0e78f83.png)

**Comment by canuse** [comment]
Some information about real time, user time, sys time:
> (i) the elapsed real time between invocation and termination,
> (ii) the user CPU time (the sum of the tms_utime and tms_cutime values in a struct tms as returned by times(2)), 
> (iii) the system CPU time (the sum of the tms_stime and tms_cstime values in a struct tms as returned by times(2)).

>struct tms {
>               clock_t tms_utime;  /\* user time \*/
>               clock_t tms_stime;  /\* system time \*/
>               clock_t tms_cutime; /\* user time of children \*/
>               clock_t tms_cstime; /\* system time of children \*/
> };

* User time contains: every thing run in user mode, like loops, branches, etc.
* System time contains: every thing that requires kernel mode, like malloc, disk io (however, this does not include the time that the program waiting for data from disk)
* not in User time and System time: CPU scheduling time (for example, running 40 processes on 20-core processor), time that waiting for data from disk/net/keyboard/...
* In short, User and Sys time are time that CPU is running the process (and its children), other time is time that the process is waiting (syn or data from device).
* Normally, `Real time = (User time +Sys time +other) / number_of_process` , and the `average CPU use = (User + Sys)/Real/number_of_process`

If Real << (User + Sys)/number_of_process, this may means that many time is spent in syn, disk io.

**Comment by canuse** [comment]
BamToH5
Always increasing when more processes are applied,
![image](https://user-images.githubusercontent.com/44812363/120328194-8124c480-c2e2-11eb-87fc-a8be7ad47992.png)
![image](https://user-images.githubusercontent.com/44812363/120328213-84b84b80-c2e2-11eb-9ad0-22da535879d3.png)
![image](https://user-images.githubusercontent.com/44812363/120328225-884bd280-c2e2-11eb-9a5f-4b99449c44c6.png)
![image](https://user-images.githubusercontent.com/44812363/120328262-900b7700-c2e2-11eb-8483-24a01f0b35e8.png)

please notice that though the sys time keeps increasing, the absolute time is small(less than 3s/process)

**Comment by canuse** [comment]
The bamToH5 contains three parts, preparing, readcounts using mclapply and write hdf5. The mclapply part can accelerate with multi core, while the preparing and write hdf5 can onl;y use limited number of cores.
![image](https://user-images.githubusercontent.com/44812363/120455435-aa019400-c38c-11eb-8cc2-5c56da00b55c.png)
![image](https://user-images.githubusercontent.com/44812363/120455450-aec64800-c38c-11eb-90a7-749eb9784df3.png)
![image](https://user-images.githubusercontent.com/44812363/120455462-b1c13880-c38c-11eb-8e38-91acbaa7e264.png)

The mclapply part is scalling good, and always increasing, the time of preparing and write hdf5 parts are similar under different number of processes. The mcapply time is still not small compared with wrtiehdf5, even in 40 cores, so maybe using as many cores as possible will be a good solution.
```
JEC21, 40core
Preprocess time : 0.221
mcapply time : 67.42
writehdf5 time : 148.145
```

Using the rprof will hugely affect the performance in multi core, and the result is very hard to see, so we can only time it manually. However, it can provide memory profile.
![image](https://user-images.githubusercontent.com/44812363/120456312-60657900-c38d-11eb-91b2-6f19f3e5648b.png)

**Comment by canuse** [comment]
In `bam_to_h5.R`, the preprocess part is short and the mcapply part is already parallelized, so we look deeper into hdf5 write part.
The hdf5 part contains a loop, each iteration will first do some calculation to get the result, then write the result to hdf5 file. The sum of execution time of calculation part and write disk part in all iterations is:
```text
JEC21, 1core
calculation: Time difference of 45.92837 secs
Write file: Time difference of 2.097302 mins
```
This part can be parallelized.

**Comment by canuse** [comment]
The hdf5 file do not support parallel write, the process that flushes the buffer later will overwrite all changes made by other processes.
A solution is to use some thing like message queue that have many producer (produces data that need to write) and one consumer (that writes into the file). https://forum.hdfgroup.org/t/hdf5-parallel-writing-to-single-file-in-different-groups-from-different-python-programms/4386
The python package `h5py` supports this, with the help of GIL and MPI, but the R package `rhdf5` do not have similar function.

**Comment by canuse** [comment]
Updated performance investigations, each figure contain 6 or 8 sub figs,
| | 1 | 2| 3|
|-|-------------|-----------------------|-----------------------------------|
| 1 | time vs core | time vs sample size | time vs gff3 size (optional) | 
| 2 | speedup rate (1/time vs core) | parallel efficiency (speedup / core) | NULL |
| 3 | memory vs core | memory vs sample size | memory vs gss_size (optional) |

**Comment by canuse** [comment]
![bamToH5](https://user-images.githubusercontent.com/44812363/122188972-69277600-ce88-11eb-964e-8abd3d9c17ed.png)
bam to h5, execution time is related to core, sample size and gff3 size

**Comment by canuse** [comment]
![cutadapt](https://user-images.githubusercontent.com/44812363/122189073-84928100-ce88-11eb-81cb-9a246ccbe34c.png)
cutadapt, execution time is related to sample size and core

**Comment by canuse** [comment]
![hisat2ORF](https://user-images.githubusercontent.com/44812363/122189137-9411ca00-ce88-11eb-980f-6d66af9cd9b4.png)
hisat2ORF, execution time is related to core, sample size and gff3 size

**Comment by canuse** [comment]
![hisat2rRNA](https://user-images.githubusercontent.com/44812363/122189220-a3911300-ce88-11eb-9f3c-85c61a00196b.png)
hisat2rrna, execution time is related to sample size and core

**Comment by canuse** [comment]
![samViewSort](https://user-images.githubusercontent.com/44812363/122189260-adb31180-ce88-11eb-985b-0af4264ed1c7.png)
samviewsort, execution time is related to core, sample size and gff3 size

**Comment by canuse** [comment]
The performance investigation can be used as a regression to estimate the time and memory consume of a new experiment.
We use the JEC21 and JdOAG1 sample the test the performance of regression model:
```
Regression test of sample JEC21
  Method: linear, Max Error(under estimate): 4.119%, Average(abs): 4.791%, Stdev 0.05
  Method: interpolate-1d, Max Error(under estimate): 5.621%, Average(abs): 4.14%, Stdev 0.046
  Method: interpolate-quad, Max Error(under estimate): 5.932%, Average(abs): 4.113%, Stdev 0.046
  Method: interpolate-cubic, Max Error(under estimate): 5.701%, Average(abs): 4.137%, Stdev 0.046
  Method: interpolate-bspine, Max Error(under estimate): 5.69%, Average(abs): 4.137%, Stdev 0.046
Regression test of sample JdAGO1
  Method: linear, Max Error(under estimate): 6.627%, Average(abs): 3.9%, Stdev 0.049
  Method: interpolate-1d, Max Error(under estimateestimate: 9.715%, Average(abs): 4.658%, Stdev 0.06
  Method: interpolate-quad, Max Error(under estimate): 10.59%, Average(abs): 4.943%, Stdev 0.064
  Method: interpolate-cubic, Max Error(under estimate): 10.751%, Average(abs): 4.999%, Stdev 0.064
  Method: interpolate-bspine, Max Error(under estimate): 10.532%, Average(abs): 4.962%, Stdev 0.064
```
Linear is simple linear regression, interpolates are interpolation based on points, see https://docs.scipy.org/doc/scipy/reference/interpolate.html

**Comment by canuse** [comment]
Update with more test on the time estimation model.
Currently the model uses linear interpolation and linear regression to estimate.
I tested all samples starts with SRR (total 36) in the example repository, each sample was executed under [1,4,8,16] cores, and measure error by `error (%)=(estimate-truth)/truth*100`. error<0 repersents the time is underestimated (the task need more time than estimates) and error>0 repersents the time is overestimated.
Underestimating is more harmful because the sge may stops our task before it finishes, while overestiamting can be neglected because the task finished earler.

In short, the cutafapt, bamToH5 and hisat2rRNA performs good, with the maximum abs(error) less than about 30%. The samViewSort and hisat2rORF are not so good, with only 2/3 error less than 30%. However, the samples are more likely to be overestimated than underestimated, which is not so harmful to the estimation model.

I also notice that the samples causes huge error are mainly less than 100 seconds, which may not cause too much damage if we manually add more time to the estimation value.

The following 5 figures are the error distrubtion.
![image](https://user-images.githubusercontent.com/44812363/123972273-9c9ef000-d9b2-11eb-95b6-0f80bd90d055.png)
![image](https://user-images.githubusercontent.com/44812363/123972282-a0327700-d9b2-11eb-9655-7a7e22be9051.png)
![image](https://user-images.githubusercontent.com/44812363/123972301-a32d6780-d9b2-11eb-841f-e25c395ea751.png)
![image](https://user-images.githubusercontent.com/44812363/123972322-a6c0ee80-d9b2-11eb-8158-0098c0f1b201.png)
![image](https://user-images.githubusercontent.com/44812363/123972351-ad4f6600-d9b2-11eb-81c4-3829a5bd04d8.png)

The following figure is the truth time of experiments cause abs error > 50%
![image](https://user-images.githubusercontent.com/44812363/123972568-e091f500-d9b2-11eb-848f-0d88e26d49d3.png)

**Comment by ewallace** [comment]
This issue is basically completed, as per @canuse’s thesis. However we don’t have an action point about what to do with this knowledge.

**Comment by canuse** [comment]
I was working on testing the performance of those tasks with different number of processes in this issue, and I think this table summarizes the figures above.
![image](https://user-images.githubusercontent.com/44812363/135367195-129c535f-c714-4f8d-a511-6a92ad5f23c2.png)

The `best number of processes` is the number of processes that have the highest speedup rate and the `efficient number of processes` is the number of processes that have a high parallel efficiency and a relatively high speedup rate.
In other words, if the user want to finish a task fast, they can config with the `best number of processes`. If their budget is limited, they can config with the `efficient number of processes`, which is slower but more cost effective.
I think this may be helpful in future versions. For example, if we allow different tasks run with different number of processes, we can configure the number of processes of each task with the value in this table.

**Comment by kavousan** [comment]
Thanks for your continued engagament, @canuse. 

Can you please explain the thinking behind the best and efficient configurations? I understand what you are saying above, I am wondering about the algorithm. Thanks!

**Comment by canuse** [comment]
I executed all combinations of different tasks, number of cores, sample sizes and gff sizes.(for example, cutadapt with 16 cores, sample size=10E8 and gff size=4000) Then I group the result by those factors and calculate the speed up and parallel efficiency. I found out that most of the speed up and parallel efficiency of the same task under different sample size and gff size are similar(the subfig 2.1 and 2.2 in the figure above). The `best number of processes` comes from the summit of speed up subfig and the `efficient number of processes` is the number of processes that have about 50% parallel efficiency.
